File: queue-reconstruction-by-height.cpp
//Time - O(n^2)
//Space - O(1)


class Solution {
public:
    
    bool static sortCompare(vector<int> & a, vector<int>& b){
        if(a[0]>b[0]) return true;
        else if(a[0]==b[0]){
            return a[1]<b[1];
        }
        return false;
    }

    

    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        vector<vector<int>> ans;
        sort(people.begin(),people.end(),sortCompare);
        
        for(auto p:people){
            ans.insert(ans.begin()+p[1],p);
        }
        
        return ans;
    }
};

File: QueueReconstructionByHeight.py
"""
Time Complexity : O(n^2) where n is the total number of people 
Space Complexity : O(n) where n is the total number of people 
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        if len(people) == 0:
            return [[]]
        result = []
        # First we dsort the elements according to their heights in descending order
        # Also in ascending order of their values for the people with same height
        people.sort(key = lambda x: (-x[0], x[1]))
        # We iterate through the sorted list and rearrage them according to their k 
        # values and return
        for i in people:
            result.insert(i[1], i)
        return result